commons-numbers-examples/examples-jmh/src/main/java/org/apache/commons/numbers/examples/jmh/core/DoubleSplitPerformance.java

```diff
@@ -0,0 +1,615 @@
+
+        /** Numbers. */
+        private double[] a;
+
+        /**
+         * Gets the factors.
+         *
+         * @return Factors.
+         */
+        public double[] getNumbers() {
+            return a;
+        }
+
+        /**
+         * Create the factors.
+         */
+        @Setup
+        public void setup() {
+            final UniformRandomProvider rng = RandomSource.create(RandomSource.XO_RO_SHI_RO_1024_PP);
+            a = new double[size];
+            for (int i = 0; i < size; i++) {
+                long bits = rng.nextLong() & SIGN_MATISSA_MASK;
+                // The exponent will either be small or big
+                if (rng.nextDouble() < edge) {
+                    bits |= EXP_SMALL;
+                } else {
+                    bits |= EXP_BIG;
+                }
+                a[i] = Double.longBitsToDouble(bits);
+            }
+        }
+    }
+
+    /**
+     * The factors to multiply.
+     */
+    @State(Scope.Benchmark)
+    public static class BiFactors {
+        /** The exponent for small numbers. */
+        private static final long EXP_SMALL = Double.doubleToRawLongBits(1.0);
+
+        /**
+         * The count of products.
+         */
+        @Param({"5000"})
+        private int size;
+
+        /**
+         * The exponent for big numbers.
+         * Two big numbers multiplied together should overflow.
+         *
+         * <p>Note: If this is set below the safe upper for Dekker's split then a method
+         * that computes the split of the two factors independently and then does the multiply
+         * may create split parts that will overflow during multiplication.
+         */
+        @Param({"600", "1000", "1023"})
+        private int exp;
+
+        /**
+         * The fraction of small numbers.
+         *
+         * <p>Note: The split numbers are used in multiplications.
+         * It is unlikely that many numbers will be larger than the upper limit.
+         * These numbers are edge cases that would cause overflow in multiplications if
+         * the other number is anywhere close to the same magnitude.
+         */
+        @Param({"1", "0.95", "0.9"})
+        private double edge;
+
+        /** Factors. */
+        private double[] a;
+
+        /**
+         * Gets the factors to be multiplied as pairs. The array length will be even.
+         *
+         * @return Factors.
+         */
+        public double[] getFactors() {
+            return a;
+        }
+
+        /**
+         * Create the factors.
+         */
+        @Setup
+        public void setup() {
+            // Validate the big exponent
+            final double d = Math.scalb(1.0, exp);
+            assert Double.isInfinite(d * d) : "Product of big numbers does not overflow";
+            final long expBig = Double.doubleToRawLongBits(d);
+
+            final UniformRandomProvider rng = RandomSource.create(RandomSource.XO_RO_SHI_RO_1024_PP);
+            a = new double[size * 2];
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).